src/server/ftpserver/options.rs

```diff
@@ -216,3 +216,42 @@ impl Default for Shutdown {
         }
     }
 }
+
+#[derive(Debug, Clone)]
+/// Variants for failed logins protection policy
+pub enum FailedLoginsPolicy {
+    /// User plus source IP address locking
+    SourceUserLock(FailedLoginsPenalty),
+    /// Source IP locking
+    SourceIPLock(FailedLoginsPenalty),
+    /// Lock the user
+    UserLock(FailedLoginsPenalty),
+}
+
+impl FailedLoginsPenalty {
+    /// Create a new FailedLoginsPenalty instance
+    pub fn new(max_attempts: u32, expires_after: Duration) -> FailedLoginsPenalty {
+        FailedLoginsPenalty { max_attempts, expires_after }
+    }
+}
+
+#[derive(Debug, Clone)]
+/// Describes the exact penalty
+pub struct FailedLoginsPenalty {
```

Comment:
Don't have to do it but as an alternative you could call this `FailedLoginsPolicy` and then embed the enum in it:

```
pub struct FailedLoginsPolicy {
    /// The maximum number of consecutive failed login attempts before the account gets locked
    pub(crate) max_attempts: u32,
    /// The expiration time since the last failed login attempt that the account gets unlocked
    pub(crate) expires_after: Duration,
    
    pub(crate) block_by: FailedLoginsBlock
}

pub enum FailedLoginsBlock {
    /// User plus source IP address locking
    IPAndUser,
    /// Source IP locking
    IP,
    /// Lock the user
    User,
}
```

Reply:
Yes good suggestion. Seems more logical